libsplinter/src/network/peer_manager/peer_map.rs

```diff
@@ -0,0 +1,383 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use super::error::PeerUpdateError;
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum PeerStatus {
+    Connected,
+    Disconnected { retry_attempts: u64 },
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct PeerMetadata {
+    pub id: String,
+    pub endpoints: Vec<String>,
+    pub active_endpoint: String,
+    pub status: PeerStatus,
+}
+
+pub struct PeerMap {
+    peers: HashMap<String, PeerMetadata>,
+    redirects: HashMap<String, String>,
+    // Endpoint to peer id
+    endpoints: HashMap<String, String>,
+}
+
+/// A map of Peer IDs to peer metadata, which also maintains a redirect table for updated peer IDs.
+///
+/// Peer metadata includes the peer_id, the list of endpoints and the current active endpoint.
+impl PeerMap {
+    pub fn new() -> Self {
+        PeerMap {
+            peers: HashMap::new(),
+            redirects: HashMap::new(),
+            endpoints: HashMap::new(),
+        }
+    }
+
+    /// Returns the current list of peer ids.
+    ///
+    /// This list does not include any of the redirected peer ids.
+    pub fn peer_ids(&self) -> Vec<String> {
+        self.peers
+            .iter()
+            .map(|(_, metadata)| metadata.id.to_string())
+            .collect()
+    }
+
+    /// Insert a new peer id and endpoints
+    pub fn insert(&mut self, peer_id: String, endpoints: Vec<String>, active_endpoint: String) {
+        let peer_metadata = PeerMetadata {
+            id: peer_id.clone(),
+            endpoints: endpoints.clone(),
+            active_endpoint,
+            status: PeerStatus::Connected,
+        };
+
+        self.peers.insert(peer_id.clone(), peer_metadata);
+
+        for endpoint in endpoints {
+            self.endpoints.insert(endpoint, peer_id.clone());
+        }
+    }
+
+    /// Remove a peer id, its endpoint and all of its redirects. Returns the active_endpoint of
+    /// the peer.
+    pub fn remove(&mut self, peer_id: &str) -> Option<String> {
+        debug!("Removing peer: {}", peer_id);
```

Comment:
This should probably be removed (higher levels, like peer manager itself should log all connection/disconnection activity.

Reply:
removed and added to the PeerManager